pallets/tinlake-investor-pool/src/lib.rs

```diff
@@ -0,0 +1,1041 @@
+					tranche
+						.value
+						.checked_add(&sol.0.mul_floor(tranche.supply))
+						.and_then(|value| value.checked_sub(&sol.1.mul_floor(tranche.redeem)))
+				})
+				.collect::<Option<Vec<_>>>()
+				.ok_or(Error::<T>::Overflow)?;
+			Self::validate_pool_constraints(
+				new_reserve,
+				pool_details.max_reserve,
+				&tranche_ratios,
+				&tranche_values,
+			)
+		}
+
+		fn validate_core_constraints(
+			currency_available: T::Balance,
+			currency_out: T::Balance,
+		) -> DispatchResult {
+			if currency_out > currency_available {
+				Err(Error::<T>::InsufficientCurrency)?
+			}
+			Ok(())
+		}
+
+		fn validate_pool_constraints(
+			reserve: T::Balance,
+			max_reserve: T::Balance,
+			tranche_ratios: &[Perquintill],
+			current_tranche_values: &[T::Balance],
+		) -> DispatchResult {
+			if tranche_ratios.len() != current_tranche_values.len() {
+				Err(Error::<T>::InvalidData)?
+			}
+
+			if reserve > max_reserve {
+				Err(Error::<T>::InsufficientReserve)?
+			}
+
+			for (count, tv) in current_tranche_values.iter().enumerate() {
+				let acc_sub_value = current_tranche_values
+					.iter()
+					.skip(count + 1)
+					.fold(Zero::zero(), |sum: T::Balance, tvs| {
+						sum.checked_add(tvs).unwrap()
+					});
+
+				let calc_sub: Perquintill = Perquintill::from_rational(acc_sub_value, *tv);
+				if calc_sub < tranche_ratios[count] {
+					Err(Error::<T>::SubordinationRatioViolated)?
+				}
+			}
+
+			Ok(())
+		}
+
+		fn do_execute_epoch(
+			pool_id: T::PoolId,
+			pool: &mut PoolDetails<T::AccountId, T::CurrencyId, T::EpochId, T::Balance>,
+			epoch: &EpochExecutionInfo<T::Balance, T::BalanceRatio>,
+			solution: &[(Perquintill, Perquintill)],
+		) -> DispatchResult {
+			pool.last_epoch_executed += One::one();
+
+			let execution: Vec<_> = epoch
+				.tranches
+				.iter()
+				.zip(solution.iter())
+				.map(|(tranche, (s_supply, s_redeem))| {
+					(
+						s_supply.mul_floor(tranche.supply),
+						s_redeem.mul_floor(tranche.redeem),
+					)
+				})
+				.collect();
+
+			let total_supply = execution
+				.iter()
+				.fold(
+					Some(Zero::zero()),
+					|acc: Option<T::Balance>, (supply, _)| {
+						acc.and_then(|acc| acc.checked_add(supply))
+					},
+				)
+				.ok_or(Error::<T>::Overflow)?;
+
+			let total_redeem = execution
+				.iter()
+				.fold(
+					Some(Zero::zero()),
+					|acc: Option<T::Balance>, (_, redeem)| {
+						acc.and_then(|acc| acc.checked_add(redeem))
+					},
+				)
+				.ok_or(Error::<T>::Overflow)?;
+
+			// Update tranche orders and add epoch solution state
+			for ((((tranche_id, tranche), solution), execution), epoch_tranche) in pool
+				.tranches
+				.iter_mut()
+				.enumerate()
+				.zip(solution.iter().copied())
+				.zip(execution.iter().copied())
+				.zip(&epoch.tranches)
+			{
+				let loc = TrancheLocator {
+					pool_id,
+					tranche_id: T::TrancheId::try_from(tranche_id)
+						.map_err(|_| Error::<T>::TrancheId)?,
+				};
+				Self::update_tranche_for_epoch(
```

Comment:
Here it probably makes sense to catch that the transfer failed and fail cleanly in one place, with the pool marked as "broken" somehow to prevent further epoch operations until we can fix it with `sudo` somehow.

That's a big enough change it probably deserves it's own PR separately from the merge work?